### PR TITLE
Fix DyanmicSparsityPattern::iterator.

### PR DESCRIPTION
**Pull request by bangerth**
In my previous commit where they were introduced, I had failed to
take into account that DynamicSparsityPattern can take an IndexSet
that denotes which rows to store. We need to skip over rows
that are not stored by this sparsity pattern.

This, unfortunately, complicated the logic significantly.

### CODE REVIEW

File: include/deal.II/lac/dynamic_sparsity_pattern.h [hunk 1]
```diff
@@ -732,19 +772,23 @@ namespace DynamicSparsityPatternIterators
...
+    // that aren't listed in a possibly passed IndexSet argument of
+    // the sparsity pattern. consequently, rather than trying to
+    // duplicate code here, just call the begin() function of the
+    // sparsity pattern itself
+    if (current_entry ==
+        ((sparsity_pattern->rowset.size()==0)
+         ?
+         sparsity_pattern->lines[current_row].entries.end()
+         :
+         sparsity_pattern->lines[sparsity_pattern->rowset.index_within_set(current_row)].entries.end()))
```
Comment (kronbichler) [inline include/deal.II/lac/dynamic_sparsity_pattern.h]: This looks clearly more expensive. Without having looked into it more deeply, would it be easier to run over the local lines only? Then hold that value as `current_row` and make a index computation for the `row()` access function. I think this would be cheaper for advance() which is the method that really matters.

**Comment by bangerth** [comment]
I thought about it, but felt that people will call `iterator->row()` quite often and so wanted to keep that fast.

How about this new version where I also store the end-of-line iterator for the current row? This makes the comparison in `advance` quite fast and we pay the penalty only every time we switch from one line to the next. I think this should be ok for most cases. What do you think?

**Comment by kronbichler** [comment]
I think this is much better. I'm curious whether the compiler will be able to optimize away the call to the index set when writing code like
```
for (DSP::iterator it=dsp.begin(row); it != dsp.end(row); ++it)
```
but I guess it should.

**Comment by kronbichler** [comment]
What I meant with my comment was whether the compiler would be able to move the construction of the end() iterator out of the loop, of course.

**Comment by bangerth** [comment]
Good question. I have no idea how to test this short of going to the assembly level. But `begin(row)` is a const function, and the accessor stores a `const` pointer to the sparsity pattern, so my best guess is that the compiler should at least in principle be able to deduce that nothing in the call to `end()` can change between different iterations of the loop.

File: include/deal.II/lac/dynamic_sparsity_pattern.h [hunk 2]
```diff
@@ -1010,15 +1070,21 @@ DynamicSparsityPattern::end (const size_type r) const
...
+         &&
+         ((row_length(row)==0)
+          ||
+          ((rowset.size() != 0) &&
+           (rowset.is_element(row) == false))))
+
+    if (row == n_rows())
+      return iterator(this);
+    else
+      return iterator(this, row, 0);
```
Comment (kronbichler) [inline include/deal.II/lac/dynamic_sparsity_pattern.h]: I'm a bit late with this comment but I realize now while building the code: Why did you change this here? I would have expected the old code to work and I must admit I don't immediately see what happens here. The `if` clause is done inside the while statement, which in turn lacks the expected `++row`. (The compiler complains about missing return arguments.)

Comment (bangerth) [inline include/deal.II/lac/dynamic_sparsity_pattern.h]: Good question. I had only later realized that `row_length()` does the right thing for rows not locally stored. But you're completely right that the `++row` is missing somewhere. Patch forthcoming -- thanks for paying more attention that I apparently did!
